### PR TITLE
[Intel MKL] Enable BF16 Softmax/SoftmaxGrad

### PR DESCRIPTION
**Pull request by Zantares**
This PR enables BF16 Softmax/SoftmaxGrad in MKLDNN backend and fixes accuracy with accumulate type in reduce operation.

* Enable BF16 Softmax with MKLDNN implementation, it implement with accumulate type inside.
* Enable BF16 SoftmaxGrad with Eigen implementation. Main modification is to implement SumReducer with accumulate type. It helps to avoid accuracy loss for BF16 reduction. The new UT can show the difference W/O accumulate type.

Backup:
Softmax with bfloat16 numeric would generally be used through auto-mixed-precision grappler pass. Default behavior of auto-mixed-precision is **NOT** to convert Softmax from float32 to bfloat16 (blacklisted for conversion). However, user could override default behavior to use bfloat16 precision of Softmax and SoftmaxGrad.

Signed-off-by: Lu Teng teng.lu@intel.com

### CODE REVIEW

**Comment by Zantares** [comment]
I separate BF16 Softmax from this PR because it has dependency with MKLDNN. We'll have another PR to submit forward part, thanks!

**Comment by mdfaijul** [comment]
@penpornk @Zantares We will update this PR soon and would like to put this on hold for a short time.
Thanks.

**Comment by gbaned** [comment]
@Zantares, @mdfaijul Any update on this PR? Please. Thanks!

**Comment by Zantares** [comment]
> @Zantares, @mdfaijul Any update on this PR? Please. Thanks!

We are trying to add new util function to have a better control of BF16 unit test. It will be updated soon in 1~2 days, thanks.

**Comment by Zantares** [comment]
Hi @gbaned @penpornk , we have fully updated this PR, now it's ready to be reviewed, thanks.
Really sorry for the delay because we need to take care of accuracy for BF16 operations.

File: tensorflow/python/ops/math_ops_test.py
Comment (penpornk) [inline tensorflow/python/ops/math_ops_test.py:55]: If the reduction is run in parallel (i.e. ThreadPoolDevice), the summation orders in both reductions could be different, causing the results to not match exactly (because of round-off errors). Could you please use `assertAllClose` instead, and pick a tolerance/threshold where the answer would match only if `f32` was used in the computation?

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR and I'm so sorry for the delay!

File: tensorflow/python/ops/math_ops_test.py
Comment (Zantares) [reply tensorflow/python/ops/math_ops_test.py:55]: Thanks for the suggestion, before modify the code I have a simple question: in original code  ThreadPoolDevice is called [here](https://github.com/tensorflow/tensorflow/pull/38630/files#diff-dce20cd1a37fce5a056d1f1c1fa02c08R75) after accumulate type conversion, so I don't think it will be rounded to low precision value in parallel. Does TF have a way to use nested ThreadPoolDevice? Or just make safe in case user use this way?

Comment (gbaned) [reply tensorflow/python/ops/math_ops_test.py:55]: @penpornk Can you please take a look on the above comment from @Zantares. Thanks!

Comment (Zantares) [reply tensorflow/python/ops/math_ops_test.py:55]: Thank you for the kindly reminding,  I ask this question is because we want to ensure accumulate type computation won't be broken in TF process, otherwise BF16 will lose accuracy. This is also the reason I use `assertAllClose` for BF16 Softmax and SoftmaxGrad  meanwhile `assertAllEqual` for `SumReduce`, we don't expect any accuracy loss happening in an operation with accumulate type inside only.

Comment (penpornk) [reply tensorflow/python/ops/math_ops_test.py:55]: Sorry for the delay! Somehow I missed @Zantares's reply the first time. Thank you for the reminder and the clarification.

My problem is with the rounding errors **within** the type f32 due to parallel computation (I will explain how it's parallelized in the next paragraph). In theory, `out_f32` in this test should have exactly the same results as the intermediate f32 results for `out_bf16` (I'll call this `out_bf16_f32_accumulator`). But since the reduction happen in parallel, there can be different roundoff errors due to different summation orders. For example, `out_f32` could be 0.9980468 and `out_bf16_f32_accumulator` could be 0.9980469, which will be 0.99609375 and 1 when cast to bfloat16, so we shouldn't use `assertAllEqual`.

Back to this reduce expression (similar to [yours](https://github.com/tensorflow/tensorflow/pull/38630/files#diff-dce20cd1a37fce5a056d1f1c1fa02c08R75) -- I just don't know how to quote a line in a PR):
https://github.com/tensorflow/tensorflow/blob/37543935526ec5c28893e448c5ee29f24b9d2aee/tensorflow/core/kernels/reduction_ops.h#L57

Eigen evaluates expressions lazily, i.e., accumulates operations through nested template typing, and only compute all operations when the expression is assigned to a device by calling the device's `TensorExecutor`. The `TensorExecutor` recursively evaluates the expression. If the device is a `ThreadPoolDevice`, evaluators can split work into shards and schedule them onto the thread pool.

`SumReducer` has an evaluator that eventually calls [Eigen::internal::FullReducer](https://gitlab.com/libeigen/eigen/-/blob/8a7f360ec3160acdb545d05712435d8ca0334357/unsupported/Eigen/CXX11/src/Tensor/TensorReduction.h#L323) that [shards the work and compute in parallel](https://gitlab.com/libeigen/eigen/-/blob/8a7f360ec3160acdb545d05712435d8ca0334357/unsupported/Eigen/CXX11/src/Tensor/TensorReduction.h#L357). 

An example call stack (most recent call on the top):
```
Eigen::internal::SumReducer<>::reduce()
Eigen::internal::InnerMostDimReducer<>::reduce()
Eigen::internal::FullReducer<>::run()
Eigen::TensorReductionEvaluatorBase<>::evalSubExprsIfNeededCommon()
Eigen::TensorReductionEvaluatorBase<>::evalSubExprsIfNeeded()
Eigen::TensorEvaluator<>::evalSubExprsIfNeeded()
Eigen::internal::TensorExecutor<>::run()
Eigen::TensorDevice<>::operator=<>()
tensorflow::functor::ReduceEigenImpl<>::operator()()
tensorflow::functor::ReduceFunctorBase<>::Reduce<>()
tensorflow::ReductionOp<>::Compute()
```

Comment (Zantares) [reply tensorflow/python/ops/math_ops_test.py:55]: Got it, so you mean the rounding error is happening inside fp32 computation which is parallelized by Eigen reducer, the accumulate type output `out_bf16_f32_accumulator` could be different with multi-running. Thanks again for the experience sharing and I will address the code soon.

Comment (Zantares) [reply tensorflow/python/ops/math_ops_test.py]: Done, replace `assertAllEqual` to `assertAllClose` with a generic tolerance `1e-3` .

**Review by penpornk — APPROVED** [review]
Thank you again for the PR!

**Comment by nammbash** [comment]
@penpornk @Zantares  I am also from Intel. Wondering why this PR is reverted after a merge?
https://github.com/tensorflow/tensorflow/commit/08b5c94d57ed1aed1120ffc0ec0a2450be61a144

Was the revert an error? or Can you hint, how we could fix and resubmit this PR as gives good performance for some of the models.

**Comment by penpornk** [comment]
@nammbash @Zantares It was automatically rolled back because it broke our internal-only tests. I have been meaning to look at it but haven't had time yet. Will update soon. Sorry for the delay!

**Comment by Zantares** [comment]
> @nammbash @Zantares It was automatically rolled back because it broke our internal-only tests. I have been meaning to look at it but haven't had time yet. Will update soon. Sorry for the delay!

What's kind of this error? Is it a crash or an accuracy test failure? Because I used accumulate type, so the accuracy will be improved for BF16 `SumReducer` and it may show different result if the internal case uses an old accuracy for comparison. Please tell us if any help we can provide, thanks!

**Comment by penpornk** [comment]
> What's kind of this error? Is it a crash or an accuracy test failure? Because I used accumulate type, so the accuracy will be improved for BF16 SumReducer and it may show different result if the internal case uses an old accuracy for comparison. Please tell us if any help we can provide, thanks!

It's an end-to-end model test. The end results differ significantly. I might not have time to dig into it before the 2.3 branch cut, so I'm thinking we can just wrap the new reducer with `INTEL_MKL && ENABLE_INTEL_MKL_BFLOAT16` for now. I'm confirming with my teammate if they are okay with this. If so, I'll bring back the changes.

**Comment by nammbash** [comment]
@penpornk Not sure I understand. If there is a problem, then wrapping it in MKL will also have the problem and test failures correct?
FYI @agramesh1 @Zantares

**Comment by penpornk** [comment]
@nammbash 

> If there is a problem, then wrapping it in MKL will also have the problem and test failures or shortcomings as opposed to fixing it. Correct?

Correct. Wrapping the reducer in `INTEL_MKL` only helps shielding the potential bug from vanilla TensorFlow, which will help this PR pass the currently failing test and be able to stay in the repo. It won't fix the potential bug in the MKL build. 

'potential' bug: There is a chance that the program works correctly but got vastly different results because of a severe error propagation in the original test result which uses just `bf16`. 

I asked @rmlarsen to help double check and he didn't see anything obviously wrong with the reducer either. We will bring back the PR with the `INTEL_MKL` guard around the reducer and will try to confirm if this is a bug or just the test's fault.

**Comment by Zantares** [comment]
@penpornk @nammbash my opinion: 

We added BF16 MklSoftmax, SoftmaxGrad and Eigen SumReducer with accumulate type. Gradients will be different if the test is using BF16 Softmax without `INTEL_MKL` because accumulate type is only for MklSoftmax rather than Eigen Softmax.
Otherwise it should be someting wrong if the failure is only appearing in SumReducer, . Please keep us posted if find anything wrong, thanks!

**Comment by penpornk** [comment]
@Zantares @nammbash Sorry for the late update! It was the test's fault and I fixed it now. I'm just waiting on a wider coverage test results to see if there's anything else failing. If things go well, I should be able to bring this PR back tomorrow.

**Pull request merged**
